stream-android-core/src/main/java/io/getstream/android/core/internal/observers/network/StreamNetworkMonitorCallback.kt

```diff
@@ -0,0 +1,221 @@
+        logger.v { "Network capabilities changed for $network" }
+        handleUpdate(network, networkCapabilities, null, UpdateReason.PROPERTIES)
+    }
+
+    override fun onLinkPropertiesChanged(network: Network, linkProperties: LinkProperties) {
+        logger.v { "Link properties changed for $network" }
+        handleUpdate(network, null, linkProperties, UpdateReason.PROPERTIES)
+    }
+
+    override fun onLost(network: Network) {
+        handleLoss(network, permanent = false)
+    }
+
+    override fun onUnavailable() {
+        handleLoss(network = null, permanent = true)
+    }
+
+    private fun resolveInitialState(): ActiveNetworkState? {
+        val defaultNetwork =
+            resolveDefaultNetwork()
+                ?: run {
+                    logger.v { "No active network available at start" }
+                    return null
+                }
+        val capabilities = connectivityManager.getNetworkCapabilities(defaultNetwork)
+        val linkProperties = connectivityManager.getLinkProperties(defaultNetwork)
+        val snapshot = buildSnapshot(defaultNetwork, capabilities, linkProperties) ?: return null
+        return ActiveNetworkState(defaultNetwork, capabilities, linkProperties, snapshot)
+    }
+
+    private fun resolveDefaultNetwork(): Network? =
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            connectivityManager.activeNetwork ?: connectivityManager.allNetworks.firstOrNull()
+        } else {
+            connectivityManager.allNetworks.firstOrNull()
+        }
+
+    private fun buildSnapshot(
+        network: Network,
+        capabilities: NetworkCapabilities?,
+        linkProperties: LinkProperties?,
+    ): StreamNetworkInfo.Snapshot? =
+        snapshotBuilder.build(network, capabilities, linkProperties).getOrElse { throwable ->
+            logger.e(throwable) { "Failed to assemble network snapshot" }
+            null
+        }
+
+    private fun handleUpdate(
+        network: Network,
+        capabilities: NetworkCapabilities?,
+        linkProperties: LinkProperties?,
+        reason: UpdateReason,
+    ) {
+        if (!shouldProcessNetwork(network)) {
+            logger.v { "[handleUpdate] Ignoring network $network; not default." }
+            return
+        }
+
+        val resolvedCapabilities =
+            capabilities ?: connectivityManager.getNetworkCapabilities(network)
+        val resolvedLink = linkProperties ?: connectivityManager.getLinkProperties(network)
+        val snapshot = buildSnapshot(network, resolvedCapabilities, resolvedLink)
+        if (snapshot == null) {
+            logger.v { "[handleUpdate] Snapshot unavailable; skipping notification." }
+            return
+        }
+
+        val newState = ActiveNetworkState(network, resolvedCapabilities, resolvedLink, snapshot)
+        val previousState = activeState.getAndSet(newState)
+
+        val networkChanged = previousState?.network != network || previousState == null
```

Comment:
[nit] The null check is already bundled in the first || operand  
```suggestion
        val networkChanged = previousState?.network != network
```